homie-controller/tests/integration.rs

```diff
@@ -0,0 +1,220 @@
+    homie.ready().await.unwrap();
+
+    // Wait until the controller knows about all required attributes of the device.
+    loop {
+        if let Some(event) = controller.poll(&mut event_loop).await.unwrap() {
+            log::trace!("Event: {:?}", event);
+            let devices = controller.devices();
+            if let Some(device) = devices.get("device_id") {
+                // For some reason we get the ready state before all the attributes of the property
+                // have been filled in, so we need to explicitly check for the unit being set.
+                if device.state == State::Ready
+                    && device.has_required_attributes()
+                    && device.nodes.len() == 1
+                    && device
+                        .nodes
+                        .get("node_id")
+                        .unwrap()
+                        .properties
+                        .get("property_id")
+                        .unwrap()
+                        .unit
+                        .is_some()
+                {
+                    break;
+                }
+            }
+        }
+    }
+
+    // Check that the device looks how we expect.
+    let devices = controller.devices();
+    let device = devices.get("device_id").unwrap();
+    log::info!("Device: {:?}", device);
+    assert_eq!(device.name, Some("Device name".to_string()));
+    assert_eq!(device.homie_version, "4.0");
+    assert_eq!(device.state, State::Ready);
+    assert_eq!(device.nodes.len(), 1);
+    let node = device.nodes.get("node_id").unwrap();
+    assert_eq!(node.name, Some("Node name".to_string()));
+    assert_eq!(node.node_type, Some("node_type".to_string()));
+    assert_eq!(node.properties.len(), 1);
+    let property = node.properties.get("property_id").unwrap();
+    assert_eq!(property.name, Some("Property name".to_string()));
+    assert_eq!(property.settable, true);
+    assert_eq!(property.unit, Some("unit".to_string()));
+    assert_eq!(property.value, None);
+
+    // Send a value from the device to the controller.
+    homie
+        .publish_value("node_id", "property_id", 42)
+        .await
+        .unwrap();
+
+    // Wait until the controller receives the value.
+    loop {
+        if let Some(event) = controller.poll(&mut event_loop).await.unwrap() {
+            log::trace!("Event: {:?}", event);
+            if let Event::PropertyValueChanged {
+                device_id,
+                node_id,
+                property_id,
+                value,
+                fresh,
+            } = event
+            {
+                assert_eq!(device_id, "device_id");
+                assert_eq!(node_id, "node_id");
+                assert_eq!(property_id, "property_id");
+                assert_eq!(value, "42");
+                assert_eq!(fresh, true);
+                break;
+            }
+        }
+    }
+
+    // Check that the device looks how we expect.
+    let devices = controller.devices();
```

Comment:
there is a lot of variable shadowing going on. Can we add some scopes to make the reader's brain work less hard?

Reply:
Done.